### PR TITLE
2229: Reduce the size of closures with `capture_disjoint_fields`

### PR DESCRIPTION
**Pull request by arora-aman**
One key observation while going over the closure size profile of rustc
was that we are disjointly capturing one or more fields starting at an
immutable reference.

Disjoint capture over immutable reference doesn't add too much value
because the fields can either be borrowed immutably or copied.

One possible edge case of the optimization is when a fields of a struct
have a longer lifetime than the structure, therefore we can't completely
get rid of all the accesses on top of sharef refs, only the rightmost
one. Here is a possible example:

```rust
struct MyStruct<'a> {
   a: &'static A,
   b: B,
   c: C<'a>,
}

fn foo<'a, 'b>(m: &'a MyStruct<'b>) -> impl FnMut() + 'static {
    let c = || drop(&*m.a.field_of_a);
    // Here we really do want to capture `*m.a` because that outlives `'static`

    // If we capture `m`, then the closure no longer outlives `'static'
    // it is constrained to `'a`
}
```

r? @nikomatsakis

### CODE REVIEW

**Comment by arora-aman** [comment]
I profile the change in closure size before and after the change:

Baseline: https://docs.google.com/spreadsheets/d/1U5vaXu490wn8Ae1LB1mSfoP_EZrGRHbUU6ADbeY8bn4/edit#gid=545238654
After optimization: 
https://docs.google.com/spreadsheets/d/1zOOeTFz4AvprEVsiPNFlkt4x_93N7WQe2uv2pX3BzXQ/edit#gid=545238654

File: src/test/ui/closures/2229_closure_analysis/optimization/edge_case_run_pass.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#![feature(capture_disjoint_fields)]
```
Comment (arora-aman) [inline src/test/ui/closures/2229_closure_analysis/optimization/edge_case_run_pass.rs:1]: waiting on the editon PR to go in for this to work

**Comment by arora-aman** [comment]
Notes from 2229 sync: https://hackmd.io/hpjgfcwaSK20uFtZibwhIw

File: src/test/ui/closures/2229_closure_analysis/optimization/edge_case_run_pass.rs (hunk 1 above)
Comment (arora-aman) [reply src/test/ui/closures/2229_closure_analysis/optimization/edge_case_run_pass.rs:1]: To change this to use edition*

**Review by nikomatsakis — APPROVED** [review]

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 2]
```diff
@@ -1967,6 +1975,44 @@ fn determine_place_ancestry_relation(
...
+///     // Here we really do want to capture `*m.a` because that outlives `'static`
+///
+///     // If we capture `m`, then the closure no longer outlives `'static'
+///     // it is constrained to `'a`
+/// }
+/// ```
+fn truncate_capture_for_optimization<'tcx>(place: &Place<'tcx>) -> Place<'tcx> {
+    let is_shared_ref = |ty: Ty<'_>| matches!(ty.kind(), ty::Ref(.., hir::Mutability::Not));
+
+    let idx = place.projections.iter().rposition(|proj| ProjectionKind::Deref == proj.kind);
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/upvar.rs:2006]:
```suggestion
    // Find the right-most deref (if any). All the projections that come after this
    // are fields or other "in-place pointer adjustments"; these refer therefore to
    // data owned by whatever pointer is being dereferenced here.
    let idx = place.projections.iter().rposition(|proj| ProjectionKind::Deref == proj.kind);
```

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 3]
```diff
@@ -1967,6 +1975,44 @@ fn determine_place_ancestry_relation(
...
+
+    match idx {
+        Some(idx) if is_shared_ref(place.ty_before_projection(idx)) => {
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/upvar.rs:2009]:
```suggestion
        // If that pointer is a shared reference, then we don't need those fields.
        Some(idx) if is_shared_ref(place.ty_before_projection(idx)) => {
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
@bors delegate+

r=me with comments added

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**